Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -794,11 +796,19 @@ extension TaskHandler: ChannelDuplexHandler {
         assert(head.version == HTTPVersion(major: 1, minor: 1),
                "Sending a request in HTTP version \(head.version) which is unsupported by the above `if`")
 
+        self.expectedBodyLength = head.headers[canonicalForm: "content-length"].first.flatMap { Int($0) }
+
         context.write(wrapOutboundOut(.head(head))).map {
             self.callOutToDelegateFireAndForget(value: head, self.delegate.didSendRequestHead)
         }.flatMap {
             self.writeBody(request: request, context: context)
         }.flatMap {
+            if let expectedBodyLength = self.expectedBodyLength, expectedBodyLength != self.actualBodyLength {
```

Comment:
At least two things missing here:

- a `hop` to the correct EventLoop
- an assertion that we are on the correct EL
- [if not already present: a test that invokes this from the wrong EL]

Reply:
ah, assertion is there, I just pasted the code above, will fix, also, we use special `callout` helper here for all EL correctness: `failTaskAndNotifyDelegate`

Reply:
@artemredkin the callout looks great but you're accessing a mutable property on `self` which can only be correct if we're on the right EL.

I kinda doubt that we're on the right EL because `context.write` returns a future from the `Channel`'s EL and `self.writeBody` returns a user-controlled future. Unsure how we can be on `taskEL` for sure.

Reply:
ahhh, you are totally correct, damn those event loops :P

Reply:
Hehe, we should probably start having a separation layer between NIO & AHC. AHC (rightly) chooses that it wants to cross ELs for performance but it's tricky to always get it right.

If we had a layer (a struct wrapping `Channel`) which would automatically `hop` all futures returned from the `Channel` operations, this should be much simpler. Then all the hopping it in one place.

Reply:
[not in this PR ofc]

Reply:
done, plus modified one of the test to callin from outside the EL